pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,509 @@
+		namespaceListerSynced: namespaceInformer.Informer().HasSynced,
+		l7Reconciler:          l7Reconciler,
+		L7PodNNDirectionMap:   l7PodNNDirMap,
+		L7SuricataSocketPath:  socketPath,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "l7VisibilityController"),
+	}
+	l7c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    l7c.addPod,
+			UpdateFunc: l7c.updatePod,
+			DeleteFunc: l7c.deletePod,
+		},
+		resyncPeriod,
+	)
+	l7c.namespaceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    l7c.addNamespace,
+			UpdateFunc: l7c.updateNamespace,
+			DeleteFunc: l7c.deleteNamespace,
+		},
+		resyncPeriod,
+	)
+	podUpdateSubscriber.Subscribe(l7c.processPodCNIUpdate)
+	return l7c
+}
+
+func (l7c *L7VisibilityController) Run(stopCh <-chan struct{}) {
+	defer l7c.queue.ShutDown()
+	klog.InfoS("Starting", "Controller", controllerName)
+	defer klog.InfoS("Shutting down", "Controller", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, l7c.podListerSynced, l7c.namespaceListerSynced) {
+		return
+	}
+	if itf, ok := l7c.interfaceStore.GetInterfaceByName(config.L7NetworkPolicyTargetPortName); ok {
+		l7c.targetPort = uint32(itf.OFPort)
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(l7c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (l7c *L7VisibilityController) worker() {
+	for l7c.processNextWorkItem() {
+	}
+}
+
+func (l7c *L7VisibilityController) processNextWorkItem() bool {
+	obj, quit := l7c.queue.Get()
+	if quit {
+		return false
+	}
+	defer l7c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		// As the item in the work queue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		l7c.queue.Forget(key)
+		klog.ErrorS(nil, "Expected string in work queue but got", "key", obj)
+		return true
+	} else if err := l7c.syncPod(key); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until
+		// another change happens.
+		l7c.queue.Forget(key)
+	} else if err == errInvalidAnnotation || err == errPodInterfaceNotFound {
+		// Handle errors
+		l7c.queue.Forget(key)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		l7c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing Podobject for L7Visibility failed, requeue", "Pod", key)
+	}
+	return true
+}
+
+func (l7c *L7VisibilityController) processPodCNIUpdate(e interface{}) {
+	podEvent := e.(types.PodUpdate)
+	podNN := k8s.NamespacedName(podEvent.PodNamespace, podEvent.PodName)
+	tcName := fmt.Sprintf("tcl7-%s", podNN)
+	if tcName == l7c.getTCNamefromPodNNDirMap(podNN) {
+		return
+	}
+	pod, err := l7c.podLister.Pods(podEvent.PodNamespace).Get(podEvent.PodName)
+	if err != nil {
+		return
+	}
```

Comment:
I am sorry as I didn't understand why is code redundant?
if a namespace already exists with annotation and we add a new pod without annotation, in this case we will be checking here for its namespace to have annotation

Reply:
@tnqn  is probably offline, so I'll take the liberty of replying. He simply means that code below is duplicated from the implementation of `namespaceAnnotationExists`, and that these lines can be replaced by a simple call to `namespaceAnnotationExists`.

*BUT*, I tend to agree with him when he writes that there is no need to consume CNI events in this controller. The NP controller is a very special case because there are security implications. You should remove `podUpdateSubscriber` from this controller, which will keep the code simpler and this PR shorter. We can always add it back later if there is a need, but it is doubtful at this stage.

Reply:
we rely on the interface created for the pod after it has received the pod IP. if we remove this CNI event case, we will skip on adding the pod with annotation due to absence of interface.

Reply:
you use a `workqueue`, so you can just log some info message with `V(2)` and requeue the item. You may want to reduce `minRetryDelay` to 1s to account for this case. I think it's better than coupling this controller with the CNI server. We can also wait for @tnqn to reply.

Reply:
Yes, the CNI event subscription is not necessary. We don't even need to requeue the item when the interface is missing. The Pod update event handler can be leveraged to enqueue the Pod efficiently. When the old Pod doesn't have IP and the new Pod has IP, it means the Pod just got its IP and interface, so it should be enqueued.

Reply:
Good point, @tushartathgur you can just ignore the CREATE / UPDATE Pod event in this case, as there will be another UPDATE event when the Pod receives its IP. No need for an error / requeue. You can still log it if you want, but with verbosity >= 4.

Reply:
I will update the code according to the below rules.
current implementation of update pod :
- new L7 related Pod annotation: 
  -  L7 related NS annotation exists: Don't add to queue (already added by namespace)
  - L7 related NS annotation doesn't exist: Add to the queue.

New Implementation:
- New IP:
  - L7 related Pod annotation exists:
    - L7 related NS annotation exists: add to queue because new interface is added
    - L7 related NS annotation doesn't exist:  add to queue because new interface is added
  - L7 related Pod annotation doesn't exist:
    - L7 related NS annotation exists: add to queue because new interface is added
    - L7 related NS annotation doesn't exist:  Don't add to queue
- No New IP:
  - L7 related Pod annotation added:
    - L7 related NS annotation exists: don't add to queue (NS annotation took care of it)
    - L7 related NS annotation doesn't exist:  add to queue 
  - L7 related Pod annotation not addedt:
    - L7 related NS annotation exists: Don't add to queue (NS annotation took care of it)
    - L7 related NS annotation doesn't exist:  Don't add to queue